### PR TITLE
Add generator checks and differentiability checks to assert_valid

### PR DESCRIPTION
**Pull request by astralcai**
The purpose of this story is to check if every operator can be differentiated properly with the parameter shift method. The gradient calculated using parameter shift is compared with backprop to verify that they are equal.

There are some cases where this does not apply:
1. Backprop does not produce the correct gradient in some cases such as for `StatePrep` and for `QubitUnitary`, as it does not take into account the constraint that the matrix must remain unitary, making this test invalid.
2. Some operator takes integers as parameters, such as `BasisState`. In this case, it does not make sense to take the gradient of integer parameters.

For these cases, a `skip_differentiation` toggle is added to `assert_valid` such that the differentiation check is skipped for these operators.

Three bugs are found as a result of adding this check. The relevant tests are xfailed:
https://github.com/PennyLaneAI/pennylane/issues/6331
https://github.com/PennyLaneAI/pennylane/issues/6333
https://github.com/PennyLaneAI/pennylane/issues/6340

Some other minor bug fixes are also included in this PR.

[sc-65197]
Fixes https://github.com/PennyLaneAI/pennylane/issues/6311

### CODE REVIEW

File: pennylane/ops/functions/assert_valid.py [hunk 1]
```diff
@@ -187,6 +187,29 @@ def _check_eigendecomposition(op):
...
+    """Checks that if an operator has a generator, it does."""
+
+    if op.has_generator:
+        gen = op.generator()
+        assert isinstance(gen, qml.operation.Operator)
+        new_op = (
+            qml.exp(gen, -1j * op.data[0])
+            if isinstance(op, qml.ops.Evolution)  # Is this consistency acceptable?
+            else qml.exp(gen, 1j * op.data[0])
+        )
```
Comment (albi3ro) [inline pennylane/ops/functions/assert_valid.py:200]: 😢 . Well, at least we are finding this.  I think we need to change `Evolution.generator` to match everywhere else... This is actually worth a bug report on its own.

Comment (astralcai) [reply pennylane/ops/functions/assert_valid.py]: I opened a bug report: https://github.com/PennyLaneAI/pennylane/issues/6311

File: pennylane/ops/functions/assert_valid.py [hunk 2]
```diff
@@ -276,6 +295,42 @@ def _check_bind_new_parameters(op):
...
+def _check_differentiation(op):
+    """Checks that the operator can be executed and differentiated correctly."""
+
+    if op.num_params == 0:
+        return
+
+    if isinstance(op, qml.ops.StatePrep):
+        # Fore StatePrep, we do not always expect parameter-shift to agree with backprop
+        # because backprop is unaware of the constraint that the matrix must remain unitary.
+        return
```
Comment (albi3ro) [inline pennylane/ops/functions/assert_valid.py:307]: Do we still need this, or should we just use `skip_differentiation=True`?

Comment (astralcai) [reply pennylane/ops/functions/assert_valid.py:307]: Let me try removing this

File: pennylane/ops/op_math/evolution.py [hunk 3]
```diff
@@ -147,7 +147,7 @@ def generator(self):
                 f"The operator coefficient {self.coeff} is not imaginary; the expected format is exp(-ixG)."
                 f"The generator is not defined."
             )
-        return self.base
+        return qml.s_prod(-1, self.base, lazy=False)
```
Comment (albi3ro) [inline pennylane/ops/op_math/evolution.py:150]: `-self.base` or `-1j*self.base` doesn't work?

File: pennylane/ops/functions/assert_valid.py (hunk 2 above)
Comment (astralcai) [reply pennylane/ops/functions/assert_valid.py:307]: I believe there were a number of templates all inheriting from `StatePrep` so I decided to skip this test for all of them all at once here.

File: pennylane/ops/op_math/evolution.py (hunk 3 above)
Comment (astralcai) [reply pennylane/ops/op_math/evolution.py:150]: Yeah I guess that works as well. Syntax highlighting didn't like it that's why I wrote it this way I believe.

**Review by albi3ro — APPROVED** [review]
Other than a changelog entry , everything looks good to me 👍

File: pennylane/ops/functions/assert_valid.py [hunk 4]
```diff
@@ -187,6 +187,25 @@ def _check_eigendecomposition(op):
         assert qml.math.allclose(decomp_mat, original_mat), failure_comment
 
 
+def _check_generator(op):
+    """Checks that if an operator has a generator, it does."""
```
Comment (PietropaoloFrisoni) [inline pennylane/ops/functions/assert_valid.py:191]: This docstring is amazing

File: pennylane/ops/functions/assert_valid.py [hunk 5]
```diff
@@ -276,6 +295,37 @@ def _check_bind_new_parameters(op):
...
+
+    if isinstance(ps, tuple):
+        for actual, expected in zip(ps, expected_bp):
+            assert qml.math.allclose(
+                actual, expected
+            ), "Parameter shift does not produce the expected Jacobian with this operator."
+    else:
+        assert qml.math.allclose(
+            ps, expected_bp
+        ), "Parameter shift does not produce the expected Jacobian with this operator."
```
Comment (PietropaoloFrisoni) [inline pennylane/ops/functions/assert_valid.py:326]: If in the future there's a case where the parameter-shift method provides the correct result (while backprop does not), should the error message be more conservative and report that there is simply a contradiction between the two results?

As you mention in the PR description, this already happened with `StatePrep` and for `QubitUnitary` (unless I am missing something)

**Review by PietropaoloFrisoni — COMMENTED** [review]
This looks great to me. Just a couple of questions

File: pennylane/ops/op_math/evolution.py [hunk 6]
```diff
@@ -147,7 +147,7 @@ def generator(self):
                 f"The operator coefficient {self.coeff} is not imaginary; the expected format is exp(-ixG)."
                 f"The generator is not defined."
             )
-        return self.base
+        return -1 * self.base
```
Comment (PietropaoloFrisoni) [inline pennylane/ops/op_math/evolution.py:151]: How is it possible that this was not causing issues before?

File: pennylane/ops/functions/assert_valid.py (hunk 4 above)
Comment (astralcai) [reply pennylane/ops/functions/assert_valid.py:191]: LOL

File: pennylane/ops/functions/assert_valid.py (hunk 5 above)
Comment (astralcai) [reply pennylane/ops/functions/assert_valid.py:326]: Makes sense, I'll update.

File: pennylane/ops/op_math/evolution.py (hunk 6 above)
Comment (astralcai) [reply pennylane/ops/op_math/evolution.py:151]: 🤷🏻‍♂️

**Review by PietropaoloFrisoni — APPROVED** [review]
Looks good!

**Pull request merged**